repos/system_upgrade/el7toel8/actors/setuprhsmonrhel8/tests/test_setuprhsmonrhel8.py

```diff
@@ -0,0 +1,134 @@
+import pytest
+
+from leapp.exceptions import StopActorExecutionError
+from leapp.libraries.actor import library
+from leapp.libraries.common import mounting, rhsm
+from leapp.libraries.stdlib import CalledProcessError, api
+from leapp.models import (TargetRHSMInfo, UsedTargetRepositories,
+                          UsedTargetRepository)
+
+
+def not_isolated_actions(raise_err=False):
+    commands_called = []
+
+    class MockNotIsolatedActions(object):
+        def __init__(self, base_dir=None):
+            pass
+
+        def call(self, cmd, **kwargs):
+            commands_called.append((cmd, kwargs))
+            if raise_err:
+                raise_call_error_instance()
+
+    return (commands_called, MockNotIsolatedActions)
+
+
+def raise_call_error_instance(args=None):
+    raise CalledProcessError(
+        message='A Leapp Command Error occured.',
+        command=args,
+        result={'signal': None, 'exit_code': 1, 'pid': 0, 'stdout': 'fake', 'stderr': 'fake'}
+    )
+
+
+class run_mocked(object):
+    def __init__(self, raise_err=False):
+        self.called = 0
+        self.args = []
+        self.raise_err = raise_err
+
+    def __call__(self, *args):
+        self.called += 1
+        self.args.append(args)
+        if self.raise_err:
+            raise_call_error_instance(args)
+
+
+class logger_mocked(object):
+    def __init__(self):
+        self.errmsg = None
+
+    def error(self, *args):
+        self.errmsg = args
+
+    def __call__(self):
+        return self
+
+
+def test_setrelease(monkeypatch):
+    commands_called, klass = not_isolated_actions()
+    monkeypatch.setattr(mounting, 'NotIsolatedActions', klass)
+    monkeypatch.setattr(api, 'consume', lambda *x: (x for x in (TargetRHSMInfo(release='6.6'),)))
+    library.set_release()
+    assert commands_called and len(commands_called) == 1
+    assert commands_called[0][0][-1] == '6.6'
+
+
+def test_setrelease_no_message(monkeypatch):
+    commands_called, klass = not_isolated_actions()
+    monkeypatch.setattr(mounting, 'NotIsolatedActions', klass)
+    monkeypatch.setattr(api, 'consume', lambda *x: (x for x in ()))
+    library.set_release()
+    assert not commands_called
+
+
+def test_setrelease_submgr_throwing_error(monkeypatch):
+    _, klass = not_isolated_actions(raise_err=True)
+    monkeypatch.setattr(mounting, 'NotIsolatedActions', klass)
+    monkeypatch.setattr(api, 'consume', lambda *x: (x for x in (TargetRHSMInfo(release='6.6'),)))
+    with pytest.raises(StopActorExecutionError):
+        library.set_release()
```

Comment:
@vinzenz, this test takes 25 seconds due to the _rhsm_retry decorator :( However I tried to monkeypatch it, or reload the module again, but without luck. Based on https://stackoverflow.com/questions/7667567/can-i-patch-a-python-decorator-before-it-wraps-a-function, I can't monkeypatch a decorator which is in the same module as the decorated function.
Do you see any solution to not wait those 25 secs?

Reply:
try this:
```python
monkeypatch.setattr(rhsm, 'set_release', rhsm.set_release.func_closure[0].cell_contents)
```

Reply:
But actually setting environment variable LEAPP_DEVEL_SKIP_RHSM to '1' should work too
monkey patch allows to set environment variables too

Reply:
The Py2 version works, thanks. BTW, are the tests on Travis run with Py3?

Reply:
The Makefile is set to create Py2.7 virtualenv and Travis runs the Makefile probably. Because when I run unit tests in my Py3 virtualenv, there's so many errors. But not in the line you suggested above - that works.

Reply:
I created a task to run unit tests in our CI with both Py2 and Py3.